Unexpected type reference between divided modules

**TypeScript Version:**  2.0.2

**Code**

sampleA.ts
```ts
{
    // local types
    type Data = string | boolean;
    let obj: Data = true;
    console.log(obj);
}
export { }
```

sampleB.ts
```
let v = "str" || true;

export { v }
```

tsconfig.json
```
{
    "compilerOptions": {
        "declaration": true,
        "skipDefaultLibCheck": true
    },
    "include": [
        "sample*.ts"
    ]
}
```
This is minimal settings.
When `"declaration": true` and `"skipDefaultLibCheck": true`, this issue is occured.

**Expected behavior:**

We can compile this code.

**Actual behavior:** 

```
$ tsc -v
Version 2.0.2
$ tsc -p ./
sampleB.ts(1,5): error TS4023: Exported variable 'v' has or is using name 'Data' from external module "/private/tmp/hogehogehoge/sampleA" but cannot be named.
```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -2150,8 +2150,13 @@ namespace ts {
                         buildSymbolDisplay(type.symbol, writer, enclosingDeclaration, SymbolFlags.Type, SymbolFormatFlags.None, nextFlags);
                     }
                     else if (!(flags & TypeFormatFlags.InTypeAlias) && type.flags & (TypeFlags.Anonymous | TypeFlags.UnionOrIntersection) && type.aliasSymbol) {
-                        const typeArguments = type.aliasTypeArguments;
-                        writeSymbolTypeReference(type.aliasSymbol, typeArguments, 0, typeArguments ? typeArguments.length : 0, nextFlags);
+                        if (type.flags & TypeFlags.Anonymous || !(flags & TypeFormatFlags.UseTypeAliasValue)) {
+                            const typeArguments = type.aliasTypeArguments;
+                            writeSymbolTypeReference(type.aliasSymbol, typeArguments, 0, typeArguments ? typeArguments.length : 0, nextFlags);
+                        }
+                        else {
+                            writeUnionOrIntersectionType(<UnionOrIntersectionType>type, nextFlags);
+                        }
                     }
                     else if (type.flags & TypeFlags.UnionOrIntersection) {
                         writeUnionOrIntersectionType(<UnionOrIntersectionType>type, nextFlags);

--- a/src/compiler/declarationEmitter.ts
+++ b/src/compiler/declarationEmitter.ts
@@ -327,7 +327,7 @@ namespace ts {
             }
             else {
                 errorNameNode = declaration.name;
-                resolver.writeTypeOfDeclaration(declaration, enclosingDeclaration, TypeFormatFlags.UseTypeOfFunction, writer);
+                resolver.writeTypeOfDeclaration(declaration, enclosingDeclaration, TypeFormatFlags.UseTypeOfFunction | TypeFormatFlags.UseTypeAliasValue, writer);
                 errorNameNode = undefined;
             }
         }

--- a/src/compiler/declarationEmitter.ts
+++ b/src/compiler/declarationEmitter.ts
@@ -341,7 +341,7 @@ namespace ts {
             }
             else {
                 errorNameNode = signature.name;
-                resolver.writeReturnTypeOfSignatureDeclaration(signature, enclosingDeclaration, TypeFormatFlags.UseTypeOfFunction, writer);
+                resolver.writeReturnTypeOfSignatureDeclaration(signature, enclosingDeclaration, TypeFormatFlags.UseTypeOfFunction | TypeFormatFlags.UseTypeAliasValue, writer);
                 errorNameNode = undefined;
             }
         }
@@ -563,7 +563,7 @@ namespace ts {
                 write(tempVarName);
                 write(": ");
                 writer.getSymbolAccessibilityDiagnostic = getDefaultExportAccessibilityDiagnostic;
-                resolver.writeTypeOfExpression(node.expression, enclosingDeclaration, TypeFormatFlags.UseTypeOfFunction, writer);
+                resolver.writeTypeOfExpression(node.expression, enclosingDeclaration, TypeFormatFlags.UseTypeOfFunction | TypeFormatFlags.UseTypeAliasValue, writer);
                 write(";");
                 writeLine();
                 write(node.isExportEquals ? "export = " : "export default ");
@@ -1025,7 +1025,7 @@ namespace ts {
                 }
                 else {
                     writer.getSymbolAccessibilityDiagnostic = getHeritageClauseVisibilityError;
-                    resolver.writeBaseConstructorTypeOfClass(<ClassLikeDeclaration>enclosingDeclaration, enclosingDeclaration, TypeFormatFlags.UseTypeOfFunction, writer);
+                    resolver.writeBaseConstructorTypeOfClass(<ClassLikeDeclaration>enclosingDeclaration, enclosingDeclaration, TypeFormatFlags.UseTypeOfFunction | TypeFormatFlags.UseTypeAliasValue, writer);
                 }
 
                 function getHeritageClauseVisibilityError(symbolAccessibilityResult: SymbolAccessibilityResult): SymbolAccessibilityDiagnostic {

--- a/src/compiler/types.ts
+++ b/src/compiler/types.ts
@@ -1946,6 +1946,7 @@ namespace ts {
         UseFullyQualifiedType           = 0x00000080,  // Write out the fully qualified type name (eg. Module.Type, instead of Type)
         InFirstTypeArgument             = 0x00000100,  // Writing first type argument of the instantiated type
         InTypeAlias                     = 0x00000200,  // Writing type in type alias declaration
+        UseTypeAliasValue               = 0x00000400,  // Writing infered type alias as its value. This is used when we emit declaration file
     }
 
     export const enum SymbolFormatFlags {

--- a/src/compiler/types.ts
+++ b/src/compiler/types.ts
@@ -1946,7 +1946,7 @@ namespace ts {
         UseFullyQualifiedType           = 0x00000080,  // Write out the fully qualified type name (eg. Module.Type, instead of Type)
         InFirstTypeArgument             = 0x00000100,  // Writing first type argument of the instantiated type
         InTypeAlias                     = 0x00000200,  // Writing type in type alias declaration
-        UseTypeAliasValue               = 0x00000400,  // Writing infered type alias as its value. This is used when we emit declaration file
+        UseTypeAliasValue               = 0x00000400,  // Serialize the type instead of using type-alias. This is needed when we emit declaration file.
     }
 
     export const enum SymbolFormatFlags {

--- a/tests/baselines/reference/declarationEmitInferedTypeAlias1.js
+++ b/tests/baselines/reference/declarationEmitInferedTypeAlias1.js
@@ -1,0 +1,30 @@
+//// [tests/cases/compiler/declarationEmitInferedTypeAlias1.ts] ////
+
+//// [0.ts]
+
+{
+    type Data = string | boolean;
+    let obj: Data = true;
+}
+export { }
+
+//// [1.ts]
+let v = "str" || true;
+export { v }
+
+//// [0.js]
+"use strict";
+{
+    var obj = true;
+}
+//// [1.js]
+"use strict";
+var v = "str" || true;
+exports.v = v;
+
+
+//// [0.d.ts]
+export {  };
+//// [1.d.ts]
+declare let v: string | boolean;
+export { v };

--- a/tests/baselines/reference/declarationEmitInferedTypeAlias1.symbols
+++ b/tests/baselines/reference/declarationEmitInferedTypeAlias1.symbols
@@ -1,0 +1,19 @@
+=== tests/cases/compiler/0.ts ===
+
+{
+    type Data = string | boolean;
+>Data : Symbol(Data, Decl(0.ts, 1, 1))
+
+    let obj: Data = true;
+>obj : Symbol(obj, Decl(0.ts, 3, 7))
+>Data : Symbol(Data, Decl(0.ts, 1, 1))
+}
+export { }
+
+=== tests/cases/compiler/1.ts ===
+let v = "str" || true;
+>v : Symbol(v, Decl(1.ts, 0, 3))
+
+export { v }
+>v : Symbol(v, Decl(1.ts, 1, 8))
+

--- a/tests/baselines/reference/declarationEmitInferedTypeAlias1.types
+++ b/tests/baselines/reference/declarationEmitInferedTypeAlias1.types
@@ -1,0 +1,23 @@
+=== tests/cases/compiler/0.ts ===
+
+{
+    type Data = string | boolean;
+>Data : Data
+
+    let obj: Data = true;
+>obj : Data
+>Data : Data
+>true : true
+}
+export { }
+
+=== tests/cases/compiler/1.ts ===
+let v = "str" || true;
+>v : Data
+>"str" || true : Data
+>"str" : string
+>true : boolean
+
+export { v }
+>v : Data
+

--- a/tests/baselines/reference/stringLiteralTypesAndTuples01.js
+++ b/tests/baselines/reference/stringLiteralTypesAndTuples01.js
@@ -38,5 +38,5 @@ function rawr(dino) {
 //// [stringLiteralTypesAndTuples01.d.ts]
 declare let hello: string, brave: string, newish: string, world: string;
 declare type RexOrRaptor = "t-rex" | "raptor";
-declare let im: "I'm", a: "a", dinosaur: RexOrRaptor;
+declare let im: "I'm", a: "a", dinosaur: "t-rex" | "raptor";
 declare function rawr(dino: RexOrRaptor): string;

--- a/tests/baselines/reference/stringLiteralTypesOverloads01.js
+++ b/tests/baselines/reference/stringLiteralTypesOverloads01.js
@@ -109,6 +109,6 @@ declare const boolean: "boolean";
 declare const stringOrNumber: "string" | "number";
 declare const stringOrBoolean: "string" | "boolean";
 declare const booleanOrNumber: "number" | "boolean";
-declare const stringOrBooleanOrNumber: PrimitiveName;
+declare const stringOrBooleanOrNumber: "string" | "number" | "boolean";
 declare namespace Consts2 {
 }

--- a/tests/cases/compiler/declarationEmitInferedTypeAlias1.ts
+++ b/tests/cases/compiler/declarationEmitInferedTypeAlias1.ts
@@ -1,0 +1,13 @@
+// @declaration: true
+// @skipDefaultLibCheck: true
+
+// @Filename: 0.ts
+{
+    type Data = string | boolean;
+    let obj: Data = true;
+}
+export { }
+
+// @Filename: 1.ts
+let v = "str" || true;
+export { v }

